### PR TITLE
vSphere Agent

### PR DESCRIPTION
**Pull request by marticus88**
this agent is in developing

### CODE REVIEW

**Comment by obino** [comment]
Running OCD test at https://ocd.appscale.com:8080/job/Daily%20Build/251 to check for regression. Thanks for the submission! Let us know when it's ready to review. Perhaps you can also create a wiki page documenting how to use and test this agent?

**Comment by marticus88** [comment]
404 error

**Comment by obino** [comment]
We haven't managed to get our internal builds available outside the intranet ... We use it as an internal reference when reviewing the code. Sorry about the confusion.

**Comment by obino** [comment]
The builds fails to pass the unit tests. You can try them typing 'rake' in the appscale directory.

**Comment by marticus88** [comment]
ok i try it

**Comment by marticus88** [comment]
you have install pysphere?

**Comment by marticus88** [comment]
You can say all the commands useful to test my code in appscale ?

**Comment by marticus88** [comment]
I correct syntax errors, now with the package installed pypshpere "pip install -U pysphere", exceeds 19 test successfully. I can do other tests.

**Comment by obino** [comment]
Great. If you need to install a new dependency, you can look into debian/appscale_install_functions.sh how to do it: we usually create a function and we have a wrapper for pip. Then add your function to appscale_install.sh.

**Comment by marticus88** [comment]
Ok.

**Comment by marticus88** [comment]
i write the pip wrapper for pysphere

File: InfrastructureManager/agents/vsphere_agent.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+[TESTING]
+Agent class for vSphere.
+The agent need pysphere 0.1.8 you can install from pip or from esayinstall tools
+
+You need setup your vCenter:
+	you need create a AppScale template with VMWare tools installed and to be powered on.
+  	you need create a head node in vSphere
+"""
+
+from agents.base_agent import BaseAgent, AgentConfigurationException, AgentRuntimeException
```
Comment (menivaitsi) [inline InfrastructureManager/agents/vsphere_agent.py]: Please put the imports in alphabetical order. Also keep it one import per line.

File: InfrastructureManager/agents/vsphere_agent.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+__author__ = 'Marco Maria Di Nicola'
+__email__ = 'marticus@hotmail.it'
+
+
+PARAM_CREDENTIALS = 'credentials'           #HOST(vCenter ip), USERNAME (vCenter), PASSWORD (vCenter)
+PARAM_IMAGE_ID = 'image_id'                 #name of AppScale Template create in vSphere some like appscale-template
+PARAM_RESOURCE_POOL = 'resource_pool'       #name of Resouce Pool in vSphere (reservate for AppScale) some like appscale-pool
+PARAM_INSTANCE_IDS = 'instance_ids'         #name of vm (ex: appscale1)
+
+class vSphereAgent(BaseAgent):
```
Comment (menivaitsi) [inline InfrastructureManager/agents/vsphere_agent.py]: Please use 2-space indentation for consistency with the rest of the code base.

File: InfrastructureManager/agents/vsphere_agent.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	"""
+	vSphere infrastructure agent class which can be used to spawn and terminate
+  	VMs in an vSphere based environment.
+	"""
+	def configure_instance_security(self, parameters):
+		"""
+		Prepare AppScale head node in vCenter and AppScale template for cloning
+		a new virtual machines
```
Comment (menivaitsi) [inline InfrastructureManager/agents/vsphere_agent.py]: Add a full stop at the end of the sentence here.

File: InfrastructureManager/agents/vsphere_agent.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+	    Returns:
+	      An instance of pysphere connection.
+		"""
+		credentials = parameters[PARAM_CREDENTIALS]
+		try:
+			server = VIServer()
+			server.connect(credentials['HOST'], credentials['USERNAME'], credentials['PASSWORD'])
+		except StandardError, e:
+			print ('Connection problem with the host {0}'.format(credentials['HOST']))
```
Comment (menivaitsi) [inline InfrastructureManager/agents/vsphere_agent.py]: Please use the logger for printing messages throughout this file, like so: https://github.com/AppScale/appscale/blob/master/InfrastructureManager/infrastructure_manager.py#L182

File: InfrastructureManager/agents/vsphere_agent.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
+"""
+[TESTING]
+Agent class for vSphere.
+The agent need pysphere 0.1.8 you can install from pip or from esayinstall tools
```
Comment (nlake44) [inline InfrastructureManager/agents/vsphere_agent.py]: esayinstall tools -> easyinstall tools

File: InfrastructureManager/agents/vsphere_agent.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+You need setup your vCenter:
+    you need create a AppScale template with VMWare tools installed and to be
```
Comment (nlake44) [inline InfrastructureManager/agents/vsphere_agent.py]: Capitalize "you" please.

File: InfrastructureManager/agents/vsphere_agent.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    powered on.
+      you need create a head node in vSphere
```
Comment (nlake44) [inline InfrastructureManager/agents/vsphere_agent.py]: Capitalize 'you' and format this docstring better. See other files as an example.

File: InfrastructureManager/agents/vsphere_agent.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+from utils import utils
+
+__author__ = 'Marco Maria Di Nicola'
+__email__ = 'marticus@hotmail.it'
+__email2__ = 'dinicola.1210022@studenti.uniroma1.it'
+
+PARAM_USERNAME = 'vusername'
+PARAM_PASSWORD = 'vpassword'
+PARAM_HOST = 'vhost'
+PARAM_RESOURCE_POOL = 'vresource_pool'       #name of Resouce Pool in vSphere (reservate for AppScale) some like appscale-pool
```
Comment (nlake44) [inline InfrastructureManager/agents/vsphere_agent.py]: Run "pylint --indent-string '  ' <filename>" on all python files. It'll complain about this line being too long.

File: InfrastructureManager/agents/vsphere_agent.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+PARAM_IMAGE_ID = 'vimage_id'                 #name of AppScale Template create in vSphere some like appscale-template
+PARAM_INSTANCE_IDS = 'instance_ids'         #name of vm (ex: appscale1)
+
+class vSphereAgent(BaseAgent):
+
```
Comment (nlake44) [inline InfrastructureManager/agents/vsphere_agent.py]: Use only two spaces when indenting.

File: InfrastructureManager/agents/vsphere_agent.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+	"""
+	vSphere infrastructure agent class which can be used to spawn and terminate
+  	VMs in an vSphere based environment.
+	"""
+	def configure_instance_security(self, parameters):
+		"""
+		Prepare AppScale head node in vCenter and AppScale template for cloning
```
Comment (nlake44) [inline InfrastructureManager/agents/vsphere_agent.py]: See other python files for correct formatting of the doc string.

File: InfrastructureManager/agents/vsphere_agent.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+		a new virtual machines
+		"""
+		return True
+
+	def open_connection(self, parameters):
+		"""
```
Comment (nlake44) [inline InfrastructureManager/agents/vsphere_agent.py]: Reformat the rest of this file.

File: debian/appscale_install.sh [hunk 12]
```diff
@@ -18,9 +18,65 @@ if [ -z "$1" ]; then
 fi
 
 case "$1" in
+<<<<<<< HEAD
```
Comment (nlake44) [inline debian/appscale_install.sh]: Look like a merge issue. This line should not be here.

File: debian/appscale_install_functions.sh [hunk 13]
```diff
@@ -18,7 +18,11 @@ fi
 VERSION_FILE="$APPSCALE_HOME_RUNTIME"/VERSION
 export APPSCALE_VERSION=$(grep AppScale "$VERSION_FILE" | sed 's/AppScale version \(.*\)/\1/')
 
+<<<<<<< HEAD
```
Comment (nlake44) [inline debian/appscale_install_functions.sh]: There are merge issues in multiple files. Please fix this and run regression testing.

**Comment by menivaitsi** [comment]
Closing this PR as inactive. @marticus88 if you're considering to keep working on this and maybe do a demo, you'd have to merge with current master first.

**Pull request closed without merging**